File: skill_sdk/tests/fixtures/c-project/src/shop/percentage_discount.c
/*
 * percentage_discount.c
 */
#include "percentage_discount.h"

static double pct_apply(const void *ctx, double total) {
    const PercentageDiscount *d = (const PercentageDiscount *)ctx;
    return total * (1.0 - d->percentage / 100.0);
}

void percentage_discount_init(PercentageDiscount *d, double pct) {
    d->percentage = pct;
}

void percentage_discount_fill(DiscountStrategy *iface, PercentageDiscount *d) {
    iface->ctx = d;
    iface->apply = pct_apply;
}

File: skill_sdk/tests/fixtures/js-project/src/shop/discount/PercentageDiscount.js
import { DiscountStrategy } from './DiscountStrategy.js';

/**
 * Applies a percentage off the raw total.
 *
 * @implements {DiscountStrategy}
 */
export class PercentageDiscount extends DiscountStrategy {
  /**
   * @param {number} percent
   */
  constructor(percent) {
    super();
    /** @type {number} */
    this._percent = percent;
  }

  /**
   * @param {import('../models/Order.js').Order} order
   * @returns {number}
   */
  apply(order) {
    return order.rawTotal() * (1.0 - this._percent / 100.0);
  }

  /** @returns {number} */
  getPercent() {
    return this._percent;
  }
}
